lib/Transforms/Scalar/DxilEraseDeadRegion.cpp

```diff
@@ -245,16 +252,32 @@ struct DxilEraseDeadRegion : public FunctionPass {
     m_DCE.EraseAndProcessOperands(Common->getTerminator());
     BranchInst::Create(BB, Common);
 
-    // Delete the region
-    for (BasicBlock *BB : Region) {
+    // Delete the region by walking post-order (starting from the predecessors of BB)
```

Comment:
Just curious, how did the original order cause problems?

Reply:
The "Region" data structure was filled out by walking the successors of the common dominator. If this code identified a chain of blocks such that the first block had instruction defs that were used in the second, it would try to delete the source defs first, which crashing because there are still pending uses. I believe deleting a region needs to be done in post order in order to satisfy those dependencies.

Reply:
Thanks for enticing me to have another look. I ended up discovering the issue was related to the "self-loop" cleanup done earlier. The shader I was chasing had a nested self-loop inside another loop that was dead, but wasn't cleaning the instructions inside of the self-loop, causing the use vectors to be wrong for the branch to the loop. This was because the self-loop was 'disconnected' in the graph, and not marked to be cleaned by the region builder.